### PR TITLE
 Optimize dispatch for fused memoryview types

### PR DESCRIPTION
**Pull request by da-woods**
By converting them to a memoryview once, checking some of the
details of that, and then using that to initialize the
Cython memoryview.

Also closes https://github.com/cython/cython/issues/3798 as an implementation detail.

I measure something like ~7% speedup on a small toy example
that selects between 3 types of memoryviews

### CODE REVIEW

**Comment by da-woods** [comment]
Draft because it's broken PyPy2

File: Cython/Compiler/FusedNode.py [hunk 1]
```diff
@@ -474,9 +479,20 @@ def _buffer_checks(self, buffer_types, pythran_types, pyx_code, decl_code, env):
         self._buffer_check_numpy_dtype(pyx_code, buffer_types, pythran_types)
         pyx_code.dedent(2)
 
-        for specialized_type in buffer_types:
-            self._buffer_parse_format_string_check(
-                    pyx_code, decl_code, specialized_type, env)
+        # creating a Cython memoryview from a Python memoryview is efficient
```
Comment (da-woods) [inline Cython/Compiler/FusedNode.py:482]: This doesn't actually seem to be the case looking in more detail (although it probably could become the case). I suspect the optimization is mainly just in being able to skip creating the Cython memoryview much of the time

File: Cython/Utility/ModuleSetupCode.c [hunk 2]
```diff
@@ -169,7 +169,7 @@
   #undef CYTHON_USE_EXC_INFO_STACK
   #define CYTHON_USE_EXC_INFO_STACK 0
   #ifndef CYTHON_UPDATE_DESCRIPTOR_DOC
-    #define CYTHON_UPDATE_DESCRIPTOR_DOC (PYPY_VERSION_HEX >= 0x07030900)
+    #define CYTHON_UPDATE_DESCRIPTOR_DOC (PYPY_VERSION_NUM >= 0x07030900)
```
Comment (mattip) [inline Cython/Utility/ModuleSetupCode.c:172]: Did this creep in from #5074, or is it on purpose?

Comment (da-woods) [reply Cython/Utility/ModuleSetupCode.c:172]: Accidentally. I was possibly trying to look at two things at the same time. I'll remove it

**Pull request merged**